ckanext/qgov/common/anti_csrf.py

```diff
@@ -104,25 +122,52 @@ def get_response_token():
     if request.environ['webob.adhoc_attrs'].has_key('response_token'):
         LOG.debug("Reusing response token from request attributes")
         token = request.response_token
-    elif request.cookies.has_key(TOKEN_FIELD_NAME) and request.cookies.has_key(TOKEN_FRESHNESS_COOKIE_NAME):
+    elif request.cookies.has_key(TOKEN_FIELD_NAME):
         LOG.debug("Obtaining token from cookie")
         token = request.cookies.get(TOKEN_FIELD_NAME)
-        if not HEX_PATTERN.match(token):
-            LOG.debug("Invalid cookie token; making new token cookie")
+        if not validate_token(token) or is_soft_expired(token):
+            LOG.debug("Invalid or expired cookie token; making new token cookie")
             token = create_response_token()
         request.response_token = token
     else:
-        LOG.debug("No fresh token found; making new token cookie")
+        LOG.debug("No valid token found; making new token cookie")
         token = create_response_token()
         request.response_token = token
 
     return token
 
+def is_soft_expired(token):
+    """Check whether the token is old enough to need rotation.
+    It may still be valid, but it's time to generate a new one.
+
+    The current rotation age is 10 minutes.
+    """
+    import time
+
+    now = int(time.time())
+    parts = token.split('!', 1)
+    message = parts[1]
+    message_parts = message.split('/', 2)
+    timestamp = int(message_parts[0])
+
+    return now - timestamp > 60 * 10
+
+def _get_secret_key():
+    from ckan.common import config
+
+    return config.get('beaker.session.secret')
+
 def create_response_token():
-    import binascii, os
-    token = binascii.hexlify(os.urandom(32))
+    import time, random, hmac, hashlib, urllib
+
+    secret_key = _get_secret_key()
+    username = urllib.quote(g.userobj.name)
```

Comment:
you should drop "/" from username if it does have one.

Reply:
Why would a username contain `/`?